x-pack/elastic-agent/pkg/core/monitoring/server/process.go

```diff
@@ -0,0 +1,195 @@
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/agent/program"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/artifact"
+	"github.com/elastic/beats/v7/x-pack/elastic-agent/pkg/core/monitoring/beats"
+)
+
+const (
+	processIDKey      = "processID"
+	monitoringSuffix  = "-monitoring"
+	separator         = "-"
+	timeout           = 10 * time.Second
+	errTypeUnexpected = "UNEXPECTED"
+
+	httpPlusPrefix = "http+"
+)
+
+var (
+	// ErrProgramNotSupported returned when requesting metrics for not supported program.
+	ErrProgramNotSupported = errors.New("specified program is not supported")
+	invalidChars           = map[rune]struct{}{
+		'"':  {},
+		'<':  {},
+		'>':  {},
+		'|':  {},
+		0:    {},
+		':':  {},
+		'*':  {},
+		'?':  {},
+		'\\': {},
+		'/':  {},
+		';':  {},
+	}
+)
+
+func processHandler() func(http.ResponseWriter, *http.Request) error {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		w.Header().Set("Content-Type", "application/json; charset=utf-8")
+
+		vars := mux.Vars(r)
+		id, found := vars[processIDKey]
+
+		if !found {
+			return errorfWithStatus(http.StatusNotFound, "productID not found")
+		}
+
+		metricsBytes, statusCode, metricsErr := processMetrics(r.Context(), id)
+		if metricsErr != nil {
+			return metricsErr
+		}
+
+		if statusCode > 0 {
+			w.WriteHeader(statusCode)
+		}
+
+		fmt.Fprint(w, string(metricsBytes))
+		return nil
+	}
+}
+
+func processMetrics(ctx context.Context, id string) ([]byte, int, error) {
+	detail, err := parseID(id)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	endpoint := beats.MonitoringEndpoint(detail.spec, artifact.DefaultConfig().OS(), detail.output)
+	if !strings.HasPrefix(endpoint, httpPlusPrefix) && !strings.HasPrefix(endpoint, "http") {
+		// add prefix for npipe and unix
+		endpoint = httpPlusPrefix + endpoint
+	}
+
+	if detail.isMonitoring {
+		endpoint += "_monitor"
+	}
+
+	hostData, err := parse.ParseURL(endpoint, "http", "", "", "stats", "")
+	if err != nil {
+		return nil, 0, errorWithStatus(http.StatusInternalServerError, err)
+	}
+
+	dialer, err := hostData.Transport.Make(timeout)
+	if err != nil {
+		return nil, 0, errorWithStatus(http.StatusInternalServerError, err)
+	}
+
+	client := http.Client{
+		Timeout: timeout,
+		Transport: &http.Transport{
+			Dial: dialer.Dial,
+		},
+	}
+
+	req, err := http.NewRequest("GET", hostData.URI, nil)
+	if err != nil {
+		return nil, 0, errorWithStatus(
+			http.StatusInternalServerError,
+			fmt.Errorf("fetching metrics failed: %v", err.Error()),
+		)
+	}
+
+	resp, err := client.Do(req.WithContext(ctx))
+	if err != nil {
+		statusCode := http.StatusInternalServerError
+		if errors.Is(err, syscall.ENOENT) {
+			statusCode = http.StatusNotFound
+		}
+		return nil, 0, errorWithStatus(statusCode, err)
+	}
+	defer resp.Body.Close()
+
+	rb, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, 0, errorWithStatus(http.StatusInternalServerError, err)
+	}
+
+	return rb, resp.StatusCode, nil
```

Comment:
Can StatusCode be != 200 here?

Reply:
i dont know, but i would rather proxy whatever is retrieved from beat than mix 200 with error message